fixtures/nesting/src/modern/withLegacyRoot.js

```diff
@@ -0,0 +1,82 @@
+import React from 'react';
+import {useContext, useMemo, useRef, useState, useLayoutEffect} from 'react';
+
+import {__RouterContext} from 'react-router';
+import ThemeContext from './shared/ThemeContext';
+
+let bridgeRecord = {
+  status: 'pending',
+  promise: null,
+  result: null,
+};
+
+export default function withLegacy(getLegacyComponent) {
+  let componentRecord = {
+    status: 'pending',
+    promise: null,
+    result: null,
+  };
+
+  return function Wrapper(props) {
+    const createLegacyRoot = readRecord(bridgeRecord, () =>
+      import('../legacy/createLegacyRoot')
+    ).default;
+    const Component = readRecord(componentRecord, getLegacyComponent).default;
+    const containerRef = useRef(null);
+    const [root, setRoot] = useState(null);
+
+    const theme = useContext(ThemeContext);
+    const router = useContext(__RouterContext);
+    const context = useMemo(
+      () => ({
+        theme,
+        router,
+      }),
+      [theme, router]
+    );
+
+    useLayoutEffect(() => {
+      if (!root) {
+        let r = createLegacyRoot(containerRef.current);
+        setRoot(r);
+        return () => r.unmount();
+      }
+    }, [createLegacyRoot, root]);
+
+    useLayoutEffect(() => {
+      if (root) {
+        root.render(Component, props, context);
+      }
+    }, [Component, root, props, context]);
+
+    return <div style={{display: 'contents'}} ref={containerRef} />;
+  };
+}
+
+function readRecord(record, createPromise) {
+  if (record.status === 'fulfilled') {
+    return record.result;
+  }
+  if (record.status === 'rejected') {
+    throw record.result;
+  }
+  if (!record.promise) {
+    record.promise = createPromise().then(
+      value => {
+        if (record.status === 'pending') {
+          record.status = 'fulfilled';
+          record.promise = null;
+          record.result = value;
+        }
+      },
+      error => {
+        if (record.status === 'pending') {
+          record.status = 'rejected';
+          record.promise = null;
+          record.result = error;
+        }
+      }
+    );
+  }
+  throw record.promise;
```

Comment:
So.. This is Suspense. I wanted to use `lazy` but that doesn't work because the component is going to be rendered by the *other* React but we want *this* React to wait. And we can't suspend in the commit phase.

I don't know if this is bad. It seems like it's part of the contract anyway and there are fairly major libs relying on it. Including Loadable Components. So we might as well use it in an example?

The benefit is that I think this pretty strongly encourages you to code-split the other React.

Reply:
We could also tuck it into a library but seems nice to have a full view of the setup for customization.